filesystem/iso9660.ksy

```diff
@@ -7,206 +7,84 @@ meta:
     pronom: fmt/468
     wikidata: Q815645
   license: CC0-1.0
+  endian: be
+  encoding: ASCII
 doc: |
   ISO9660 is standard filesystem used on read-only optical discs
   (mostly CD-ROM). The standard was based on earlier High Sierra
   Format (HSF), proposed for CD-ROMs in 1985, and, after several
-  revisions, it was accepted as ISO9960:1998.
-
+  revisions, it was accepted as ISO9660:1998.
   The format emphasizes portability (thus having pretty minimal
   features and very conservative file names standards) and sequential
   access (which favors disc devices with relatively slow rotation
   speed).
+doc-ref:
+  - ecma-119 https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-119.pdf
+  - susp http://www.ymi.com/ymi/sites/default/files/pdf/Systems%20Use%20P1281.pdf
+  - rrip http://www.ymi.com/ymi/sites/default/files/pdf/Rockridge.pdf
+  - rras http://www.estamos.de/makecd/Rock_Ridge_Amiga_Specific
+  - rrzf https://dev.lovelyhq.com/libburnia/web/wikis/Zisofs
+instances:
+  sector_size:
+    doc-ref: ecma-119 6.1.2
+    value: 0x800
+  volume_descriptor_set:
+    io: _root._io
+    pos: _root.sector_size * 0x10
+    type: volume_descriptor
+    size: sector_size
+    repeat: until
+    repeat-until: _.type == volume_type::volume_descriptor_set_terminator
+enums:
+  volume_type:
+    0x00: boot_record_volume_descriptor
+    0x01: primary_volume_descriptor
+    0x02: supplementary_volume_descriptor
+    0x03: volume_partition_descriptor
+    0xff: volume_descriptor_set_terminator
 types:
-  vol_desc:
-    seq:
-      - id: type
-        type: u1
-      - id: magic
-        contents: "CD001"
-      - id: version
-        type: u1
-      - id: vol_desc_boot_record
-        type: vol_desc_boot_record
-        if: type == 0
-      - id: vol_desc_primary
-        type: vol_desc_primary
-        if: type == 1
-  vol_desc_boot_record:
+  u2bi:
+    doc-ref: ecma-119 7.2.3
     seq:
-      - id: boot_system_id
-        type: str
-        size: 32
-        encoding: UTF-8
-      - id: boot_id
-        type: str
-        size: 32
-        encoding: UTF-8
-  vol_desc_primary:
-    doc-ref: 'http://wiki.osdev.org/ISO_9660#The_Primary_Volume_Descriptor'
+      - id: le
+        type: u2le
+      - id: be
+        type: u2be
```

Comment:
```yaml
valid: le
```

Reply:
I don't understand your comment.
I expect both fields to have a 16 bit number, one is le and the other is be encoded

Reply:
Must the 2 numbers be the same?

Reply:
Technically yes, but the le = le ordered and the be = be ordered.
If you look at it as a string you will see the 2 bytes are swapped.

Reply:
```yaml
valid: le
```

is equivalent to

```yaml
valid:
  eq: le
```

Reply:
The spec was designed to not burden you with le <-> be conversion.
So compairing these values kind of feels like negating the original intent.

On a rational note... why did they ever create extra IO on a slow medium while CPU's can do this so much faster?

Reply:
> The spec was designed to not burden you with le <-> be conversion.
> So compairing these values kind of feels like negating the original intent.

I don't think so - you already cause the le <-> be conversion to occur when you use `u2le` and `u2be`. It's not that they are not parsed until you access their value: `seq` attributes are parsed immediately (you'd need to use `instances` which are parsed only on demand - then it would make sense). To avoid byte swaps entirely, you need to stop using integers with non-native endianness. However, native endianness can't be determined natively using the KS language, so you have to determine externally and pass it into the KSY via [`params`](https://doc.kaitai.io/user_guide.html#param-types). Some KS runtime libraries even provide this information, so you can use it:

[**kaitai_struct_cpp_stl_runtime** / kaitai/kaitaistream.cpp:9](https://github.com/kaitai-io/kaitai_struct_cpp_stl_runtime/blob/4da1668426bd67aaf1932628c1c7db76ff77665c/kaitai/kaitaistream.cpp#L9)

```cpp
#define __BYTE_ORDER    BYTE_ORDER
```

[**kaitai_struct_csharp_runtime** / KaitaiStream.cs:39](https://github.com/kaitai-io/kaitai_struct_csharp_runtime/blob/ac9add4de3f5563471981e078f345c174293bf49/KaitaiStream.cs#L39)

```csharp
static readonly bool IsLittleEndian = BitConverter.IsLittleEndian;
```

[**kaitai_struct_javascript_runtime** / KaitaiStream.js:493](https://github.com/kaitai-io/kaitai_struct_javascript_runtime/blob/2de2c4393182b91c344bfadad0583f3694decd48/KaitaiStream.js#L493)

```js
KaitaiStream.endianness = new Int8Array(new Int16Array([1]).buffer)[0] > 0;
```

---

I don't think that detecting native endianness to avoid endian conversions is necessarily worth it, it depends on the use case. However, if you decide to keep the `u2be` + `u2le` pair in `seq`, I guess that checking their values with `valid/eq` (kaitai-io/kaitai_struct#435) makes sense, because you want to ensure that they store the same values.

Reply:
Added